python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,294 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Utility module for various conversions around PrimitiveType implementations
+
+This module enables:
+    - Converting partition strings to built-in python objects
+    - Converting a value to a byte buffer
+    - Converting a byte buffer to a value
+
+Note:
+    Conversion logic varies based on the PrimitiveType implementation. Therefore conversion functions
+    are defined here as generic functions using the @singledispatch decorator. For each PrimitiveType
+    implementation, a concrete function is registered for each generic conversion function. For PrimitiveType
+    implementations that share the same conversion logic, registrations can be stacked.
+"""
+import struct
+import uuid
+from decimal import Decimal, localcontext
+from functools import singledispatch
+from typing import Union
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    PrimitiveType,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+
+def convert_decimal_to_unscaled_value(decimal_type_var: DecimalType, value: Decimal) -> int:
+    """Get an unscaled value given a Decimal value and a DecimalType instance with defined precision and scale
+
+    Args:
+        decimal_type_var (DecimalType): A DecimalType instance with precision and scale
+        value (Decimal): A Decimal instance
+
+    Raises:
+        ValueError: If either the scale or precision of `value` does not match that defined in the DecimalType instance provided as `decimal_type_var`
+
+    Returns:
+        int: The unscaled value
+    """
+    value_as_tuple = value.as_tuple()
+    value_precision = len(value_as_tuple.digits)
+    value_scale = -value_as_tuple.exponent
+
+    if value_scale != decimal_type_var.scale:
+        raise ValueError(f"Cannot serialize value, scale of value does not match type {decimal_type_var}: {value_scale}")
+    elif value_precision > decimal_type_var.precision:
+        raise ValueError(
+            f"Cannot serialize value, precision of value is greater than precision of type {decimal_type_var}: {value_precision}"
+        )
+
+    with localcontext() as ctx:
+        ctx.prec = decimal_type_var.precision
+        sign, digits, exponent = value.as_tuple()
+        value_w_adjusted_scale = Decimal((sign, digits, exponent + decimal_type_var.scale))
```

Comment:
When converting using the direct tuple construction, do we need to adjust the local context? According to [this intro to Decimal page](https://www.pythontutorial.net/advanced-python/python-decimal/), "Notice that the decimal context precision only affects the arithmetic operation, not the Decimal constructor".

That means we should just be able to return the adjusted value with scale 0.

Also, since we don't need to set the context precision, we don't need the decimal type at all. In fact, it doesn't make much sense to check the value against the type here either. I think this method could be focused on just getting the unscaled value and the checks can happen in the caller, if needed.

Reply:
Oh wow you're absolutely right. I made those changes and it's much cleaner now.